synapse/dendrite/jobs.py

```diff
@@ -0,0 +1,214 @@
+import json
+import time
+import logging
+import collections
+import synapse.common as s_common
+import synapse.cortex as s_cortex
+import synapse.eventbus as s_eventbus
+import synapse.telepath as s_telepath
+
+logger = logging.getLogger(__name__)
+
+class Jobs(s_eventbus.EventBus):
+    '''
+    Jobs provides a simple remote queue for managing Parser jobs in Dendrite.
+
+    Parser jobs are added to named queues, an event is fired on the Service Bus, the Parsers registered for
+    the named queue receive the event and proceed to retrieve the job for processing. Basic job lifecycle management
+    is provided where a job can be in one of the following states: queued, working, and completed.
+
+    Args:
+        core (str):
+            URL of the Cortex to be used for internal management of the job queues.
+
+        svcbus (str):
+            URL of the SvcBus that events will be emitted on.
+    '''
+    def __init__(self, core, svcbus):
+        s_eventbus.EventBus.__init__(self)
+        self.currMs = None
+        self.jobNum = None
+        self.core = s_cortex.openurl(core)
+        self.core.addTufoForm('dendrite:job', ptype='guid')
+        self.core.addTufoProp('dendrite:job', 'queue', ptype='str', req=True)
+        self.core.addTufoProp('dendrite:job', 'data', ptype='str', req=True)
+        self.core.addTufoProp('dendrite:job', 'runkey', ptype='int', req=True)
+        self.core.addTufoProp('dendrite:job', 'status', ptype='str', req=True)
+        self.core.addTufoProp('dendrite:job', 'failed_at', ptype='int', req=True)
+        self.core.addTufoProp('dendrite:job', 'queued_at', ptype='int', req=True)
+        self.core.addTufoProp('dendrite:job', 'working_at', ptype='int', req=True)
+        self.core.addTufoProp('dendrite:job', 'completed_at', ptype='int', req=True)
+        self.svcbus = s_telepath.openurl(svcbus)
+
+    def put(self, queue, job):
+        '''
+        Adds job to a queue.
+
+        The job is added to the queue and event of the form ``'dendrite:jobs:%s' % queue`` is fired on the Service Bus.
+
+        Args:
+            queue (str):
+                Name of the queue to add the job to.
+
+            job (dict):
+                Dict containing any properties needed to represent the job.
+        '''
+        logger.debug('Adding job %s to queue %s', job, queue)
+        props = self._buildJobTufo(queue, job)
+        self.core.formTufoByProp('dendrite:job', s_common.guid(), **props)
+        self.svcbus.fire(self._eventName(queue), job=job, queue=queue)
+
+    def get(self, queue):
+        '''
+        Get the next job off the queue specified.
+
+        Args:
+            queue (str):
+                The name of the queue to retrieve a job from.
+
+        Returns:
+            dict: The next job if one exists, None otherwise.
+        '''
+        tufos = sorted(self._jobsByQueue(queue), key=lambda t: t[1].get('dendrite:job:runkey'))
```

Comment:
I hear you! I was trying to find a way to get a 'order by runkey limit 1'. What are your thoughts about introducing a cortex API to specify order? That would allow certain implementations to optimize appropriately.